lib/icinga/dependency.hpp

```diff
@@ -42,21 +50,220 @@ class Dependency final : public ObjectImpl<Dependency>
 
 private:
 	Checkable::Ptr m_Parent;
 	Checkable::Ptr m_Child;
 
-	static bool m_AssertNoCyclesForIndividualDeps;
+	// Whether we've already performed the cyclic reference check for all dependencies on Icinga 2 startup.
+	static bool m_CyclicReferenceForAllDependenciesAsserted;
 
 	static bool EvaluateApplyRuleInstance(const Checkable::Ptr& checkable, const String& name, ScriptFrame& frame, const ApplyRule& rule, bool skipFilter);
 	static bool EvaluateApplyRule(const Checkable::Ptr& checkable, const ApplyRule& rule, bool skipFilter = false);
 };
 
+/**
+ * A DependencyGroup represents a set of dependencies that are somehow related to each other.
+ *
+ * The main purpose of this class is grouping all dependency objects that are related to each other in some way.
+ * More specifically, let's say we have a dependency graph like this:
+ * @verbatim
+ *                      PP1              PP2
+ *                      /\               /\
+ *                      ||               ||
+ *                  ----||–––––––––––––––||–––––
+ *                      P1 - ( "RG1" ) - P2
+ *                  ----––––––––––––––––––––––––
+ *                          /\      /\
+ *                          ||      ||
+ *                          C1      C2
+ * @endverbatim
+ * The arrows represent a dependency relationship from bottom to top, i.e. both "C1" and "C2" depend on
+ * their "RG1" redundancy group, and "P1" and "P2" depend each on their respective parents (PP1, PP2 - no group).
+ * Now, as one can see, both "C1" and "C2" have identical dependencies, that is, they both depend on the same
+ * redundancy group "RG1" (these could e.g. be constructed through some Apply Rules).
+ *
+ * So, instead of having to maintain two separate copies of that graph, we can bring that imaginary redundancy group
+ * into reality by putting both "P1" and "P2" into an actual DependencyGroup object. However, we don't really put "P1"
+ * and "P2" objects into that group, but rather the actual Dependency objects of both child Checkables. Therefore, the
+ * group wouldn't just contain 2 members, but 4 members in total, i.e. 2 for each child Checkable (C1 -> {P1, P2} and
+ * C2 -> {P1, P2}). This way, both child Checkables can just refer to that very same DependencyGroup object.
+ *
+ * However, since not all dependencies are part of a redundancy group, we also have to consider the case where
+ * a Checkable has dependencies that are not part of any redundancy group, like P1 -> PP1. In such situations,
+ * each of the child Checkables (e.g. P1, P2) will have their own unique non-shared DependencyGroup object.
+ * This allows us to keep the implementation simple and treat redundant and non-redundant dependencies in the same
+ * way, without having to introduce any special cases everywhere. So, in the end, we'd have 3 dependency groups in
+ * total, i.e. one for the redundancy group "RG1" (shared by C1 and C2), and two distinct groups for P1 and P2.
+ *
+ * @ingroup icinga
+ */
```

Comment:
What would happen if C1 had two more dependencies to new checkables P3 and P4 without any redundancy group? If I understand your explanation correctly, that would create a single new `DependencyGroup` that contains both new C1->P3 and C1->P4 dependencies. If so, I find it surprising that dependencies inside a redundancy are grouped by parent but those outside of any redundancy group are grouped by child.

Reply:
> If I understand your explanation correctly, that would create a single new `DependencyGroup` that contains both new C1->P3 and C1->P4 dependencies.

Exactly 👍!

> If so, I find it surprising that dependencies inside a redundancy are grouped by parent but those outside of any redundancy group are grouped by child.

Please read the documentation for the `CompositeKeyType` and `MemberValueType` type aliases to understand why!